// @defi-wonderland/smock/dist/src/observable-vm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObservableVM = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
class ObservableVM {
    constructor(vm) {
        if (!vm)
            throw new Error('VM is not defined');
        this.vm = vm;
        this.beforeTx$ = ObservableVM.fromEvent(vm, 'beforeTx');
        this.beforeMessage$ = ObservableVM.fromEvent(vm, 'beforeMessage');
        this.afterMessage$ = ObservableVM.fromEvent(vm, 'afterMessage');
    }
    getManager() {
        return (this.vm.pStateManager || this.vm.stateManager);
    }
    getBeforeTx() {
        return this.beforeTx$;
    }
    getBeforeMessages() {
        return this.beforeMessage$.pipe(operators_1.filter((message) => !!message.to));
    }
    getAfterMessages() {
        return this.afterMessage$;
    }
    static fromEvent(vm, eventName) {
        const subject = new rxjs_1.Subject();
        vm.on(eventName, (event) => subject.next(event));
        return subject.asObservable().pipe(operators_1.share());
    }
}
exports.ObservableVM = ObservableVM;
